packages/react-keyring/src/Authenticator.tsx

```diff
@@ -0,0 +1,132 @@
+import React, {
+  useState, createContext, useContext, useCallback, useMemo
+} from 'react'
+import { useKeyring, KeyringContextState, KeyringContextActions } from './providers/Keyring'
+
+export type AuthenticatorContextState = KeyringContextState & {
+  /**
+   * email to be used to "log in"
+   */
+  email?: string
+  /**
+   * has the authentication form been submitted?
+   */
+  submitted: boolean
+  /**
+   * A callback that can be passed to an `onSubmit` handler to
+   * register a new space or log in using `email`
+   */
+  handleRegisterSubmit?: (e: React.FormEvent<HTMLFormElement>) => Promise<void>
+}
+
+export type AuthenticatorContextActions = KeyringContextActions & {
+  /**
+   * Set an email to be used to log in or register.
+   */
+  setEmail: React.Dispatch<React.SetStateAction<string>>
+}
+
+export type AuthenticatorContextValue = [
+  state: AuthenticatorContextState,
+  actions: AuthenticatorContextActions
+]
+
+export const AuthenticatorContext = createContext<AuthenticatorContextValue>([
+  {
+    spaces: [],
+    submitted: false
+  },
+  {
+    setEmail: () => { throw new Error('missing set email function') },
+    loadAgent: async () => { },
+    unloadAgent: async () => { },
+    resetAgent: async () => { },
+    createSpace: async () => { throw new Error('missing keyring context provider') },
+    setCurrentSpace: async () => { },
+    registerSpace: async () => { },
+    cancelRegisterSpace: () => { },
+    getProofs: async () => []
+  }
+])
+
+/**
+ * Top level component of the headless Authenticator.
+ *
+ * Must be used inside a KeyringProvider.
+ *
+ * Designed to be used by Authenticator.Form, Authenticator.EmailInput
+ * and others to make it easy to implement authentication UI.
+ */
+export function Authenticator (props: any): JSX.Element {
+  const [state, actions] = useKeyring()
+  const { createSpace, registerSpace } = actions
+  const [email, setEmail] = useState('')
+  const [submitted, setSubmitted] = useState(false)
+
+  const handleRegisterSubmit = useCallback(async (e: React.FormEvent<HTMLFormElement>) => {
+    e.preventDefault()
+    setSubmitted(true)
+    try {
+      await createSpace()
+      await registerSpace(email)
+    } catch (err: any) {
+      throw new Error('failed to register')
```

Comment:
Do not smother!

```suggestion
      throw new Error('failed to register', { cause: err })
```

Separately, we should probably not throw this but instead expose an error state variable. This function will likely be passed directly to the `onSubmit` handler for the form so there's no opportunity to catch it. Can be a follow up ;)

Reply:
ah great call - I think TypeScript complained when I tried to pass `cause` previously, but I'll take another crack at it.

totally agree re: not throwing this, I'll get that into a followup!